File: C++/MergeSort.cpp
#include <iostream>
#include <chrono>

using namespace std;
using namespace std::chrono;

void swap(int &x, int &y) {
    int temp = x;
    x = y;
    y = temp;
}

// Merge Sort Function
void merge(int a[], int left, int mid, int right) {
    int i = 0, j = 0;
    int lSize = mid - left + 1;
    int rSize = right - mid;
    
    int lArr[lSize], rArr[rSize];
    
    for(i = 0; i < lSize; i++)
       lArr[i] = a[left+i];
    for(j = 0; j < rSize; j++)
       rArr[j] = a[mid+1+j];
    
    i = 0; j = 0;
    int k = left;
    while(i < lSize && j < rSize) {
       if(lArr[i] <= rArr[j]) {
          a[k] = lArr[i];
          i++;
       }
       else {
          a[k] = rArr[j];
          j++;
       }
       k++;
    }
    while(i < lSize) {
       a[k] = lArr[i];
       i++; k++;
    }
    while(j < rSize) {
       a[k] = rArr[j];
       j++; k++;
    }
}
void mergeSort(int arr[], int left, int right) {
    int mid;
    if(left < right) {
        mid = left + (right - left)/2;

        mergeSort(arr, left, mid);
        mergeSort(arr, mid+1, right);
        merge(arr, left, mid, right);
    }
}


void printArray(int arr[], int n) {
    cout<<"\n Array : ";
    for(int i = 0; i < n; i++)
        cout<<arr[i]<<" ";
}

int main() {
    //Input size
    int n;
    cin>>n;
    
    int arr[n];
    for(int i = 0; i < n; i++)
        cin>>arr[i];
    
    auto start = high_resolution_clock::now();
    mergeSort(arr, 0, n-1);
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(end-start);
    
    printArray(arr, n);
    cout<<"\n Time Used : "<<duration.count()<<"microsecond \n";
    
    return 0;
}

File: Java/MergeSort.java
import java.util.Scanner;

class MergeSort {

    void merge(int arr[], int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
  
        int L[] = new int[n1];
        int R[] = new int[n2];
  
        for (int i = 0; i < n1; ++i)
            L[i] = arr[left + i];
        for (int j = 0; j < n2; ++j)
            R[j] = arr[mid + 1 + j];
   
	int i = 0, j = 0;
  
        int k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            }
            else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }
  
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }
  
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }
  
    void mergeSort(int arr[], int left, int right) {
        if (left < right) {
            int mid = left + (right-left)/2;

            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }

    // Array printing function
    void printArray(int arr[], int n) {

	for(int i = 0; i < n; i++)
	    System.out.print(arr[i] + " ");
	System.out.println();
    }

    //Driver function
    public static void main(String args[]) {
	Scanner sc = new Scanner(System.in);
	MergeSort obj = new MergeSort();

	int n;	
	n = sc.nextInt();	
	int arr[] = new int[n];
	for(int i = 0; i < n; i++)
	    arr[i] = sc.nextInt();

	obj.mergeSort(arr, 0, n-1);
	System.out.println("Sorted Array:");
	obj.printArray(arr, n);
    }
}
